### PR TITLE
Single linked list

### PR DESCRIPTION
**Pull request by arccth**

### CODE REVIEW

File: course4/homework-list/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
+
+set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+add_executable(list_homework
+	test.cc
+        list.hh)
```
Comment (maciekgajewski) [inline course4/homework-list/CMakeLists.txt]: something wrong with the white-space here

File: course4/homework-list/list.hh [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        return current -> isTerminal != other.current -> isTerminal;
+    }
+};
+
+template<typename T>
+class List
+{
+private:
+    int mSize = 0;
+    std::shared_ptr<ListElement<T>> terminalElement = ListElement<T>::GetTerminalElement();
```
Comment (maciekgajewski) [inline course4/homework-list/list.hh]: This makes the default constructor expensive, which goes agatis good practices.

File: course4/homework-list/list.hh [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        return mSize;
+    }
+
+    T& front() const
+    {
+        return head->item;
+    }
+
+    void push_front(T newItem)
```
Comment (maciekgajewski) [inline course4/homework-list/list.hh]: pass newItem by cosnt reference, to avoid unnecessary copying.

File: course4/homework-list/list.hh [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        std::shared_ptr<ListElement<T>> oldHead = std::move(head);
+        head = std::make_shared<ListElement<T>>(newItem);
+        mSize++;
+        head->next = std::move(oldHead);
+    }
+
+    T pop_front()
```
Comment (maciekgajewski) [inline course4/homework-list/list.hh]: Is the signature of std::forward_list::pop_front the same?

File: course4/homework-list/test.cc [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    list.push_front(2);
+    list.push_front(3);
+    list.push_front(2);
+    list.push_front(1);
+    list.push_front(2);
+
+    int count = std::count_if(list.begin(), list.end(), [](int x) {return x == 2;});
+    assert(count == 3);
+}
+
```
Comment (maciekgajewski) [inline course4/homework-list/test.cc:49]: Please, wrtie the followign algorithm, and thest both containers on it:

"count_until" - an algorithm that counts the occurences of a value, nween the beginning of a range, and the first occurence of some "stop" element.
If there is no "stop" element, then the count should encompass the entire container.

Examples:

```
list = {1, 2, 1, ,2, 3, 1, ,2 };
auto count1 = count_until(list.begin(), list.end(), 1/*search*/, 3/*stop*/);
assert(count1 == 2);
auto count2 = count_until(list.begin(), list.end(), 1/*search*/, 7/*stop*/);
assert(count2 == 3);
```

**Review by maciekgajewski — COMMENTED** [review]
Good.
Bonus task: Please, implement the algorithm as described in a comment, expend the test to use it.
Bonus question: what is the performance impact of the shared_ptr? What are the benefits of using shared_ptr?

File: course4/homework-list/CMakeLists.txt (hunk 1 above)
Comment (arccth) [reply course4/homework-list/CMakeLists.txt]: Done

File: course4/homework-list/list.hh (hunk 3 above)
Comment (arccth) [reply course4/homework-list/list.hh]: Done

File: course4/homework-list/list.hh (hunk 4 above)
Comment (arccth) [reply course4/homework-list/list.hh]: Changed to void

File: course4/homework-list/test.cc (hunk 5 above)
Comment (arccth) [reply course4/homework-list/test.cc:49]: Done

File: course4/homework-list/list.hh (hunk 2 above)
Comment (arccth) [reply course4/homework-list/list.hh]: Done

**Pull request closed without merging**